ipbench2/doc/ipbenchd.man
.TH "ipbenchd" "1" "Mar 2024" "Ian Wienand, Lesley Rossouw, Peter Chubb, Christopher Irving" ""
.SH "NAME"
.LP
ipbenchd \- Client daemon for the ipbench testing suite
.SH "SYNTAX"
.LP
ipbenchd [\fB\-i\fR <\fIip\fP>|\fB\-\-ip\fP=<\fIip\fP>]
[\fB\-p\fR <\fIport\fP>|\fB\-\-port\fP=<\fIport\fP>]
[\fB\-t\fP|\fB\-\-target\fP] [\fB\-d\fP|\fB\-\-debug\fP]
.br
ipbenchd \fB\-h\fP|\fB\-\-help\fR
.br
ipbenchd \fB\-\-version\fP
.SH "DESCRIPTION"
.LP
This program is the client portion of the ipbench testing suite.  Each client must be running this daemon to receive commands from the controlling node. The daemon must be run as root to allow access to ports. It also does not fork into a separate process automatically. So you may want to run something like
.LP
.B sudo /usr/bin/ipbenchd &
.LP
to run the daemon in the background.
.br
This daemon can also be used as the target portion of the testing suite, by running it with \fB\-t\fR
.LP
For more on the structure of the ipbench testing suite, see the man page for ipbench(1)
.SH "OPTIONS"
.LP
.TP
\fB\-\-version\fR
Show the program's version number and exit.
.TP
\fB\-h\fP|\fB\-\-help\fR
Output help information.
.TP
\fB\-i\fR <\fIip\fP>|\fB\-\-ip\fP=<\fIip\fP>
Makes the daemon bind to the specified \fIip\fP. Default is to bind to all available ip addresses.
.TP
\fB\-p\fR <\fIport\fP>|\fB\-\-port\fP=<\fIport\fP>
Makes the daemon listen on the specified \fIport\fP.
.TP
\fB\-d\fP|\fB\-\-debug\fR
Enable debugging; provide verbose output.
.TP
\fB\-t\fP|\fB\-\-target\fR
Put the daemon in target mode
.SH "COMMUNICATING WITH IPBENCHD"
.LP
By sending messages to the port ipbenchd is listening on (e.g. using telnet) you
can issue commands to ipbenchd. The available commands are listed here.

These first four commands can be issued at any time.

.TP
.B
HELP
.br
Prints a list of all ipbenchd commands.
.TP
.B
STATUS
.br
Show the status of ipbenchd.
.TP
.B
QUIT
.br
Close the connection to ipbenchd.
.TP
.B
ABORT
.br
If ipbenchd is currently running a test, end the test immediately without
processing data and return to ready state so another test can be configured and run.
Then close the connection to ipbenchd.
This command is useful if a previous test failed unexpectedly and ipbenchd
has been left unable to accept new tests.
.LP
These remaining commands
take ipbenchd through the process of preparing, running, and completing
a test. Ipbenchd will only accept one of these commands if it has been brought
to the correct state by successful runs of the previous commands in order.
If not, a syntax error will be returned.
.TP
.B
HELLO
.br
Prepare ipbenchd to run a new test. This command can only be run when ipbenchd
is in its initial ready state and not running a test.
.TP
.B
LOAD <test-name>
.br
Specify which test will be run. This command can only be run after HELLO.
.TP
.B
SETUP
.br
Configure how the test will be run. This command can only be run after LOAD.
There are two versions of the setup command, based on whether ipbenchd is
running in client mode or target mode. In client mode, SETUP must be run as
.br

\fBSETUP\fP target::<test-target>||port::<test-port>||args::<test-args>
.br

This passes the target, port on the target, and arguments for the
test in question to use. In target mode, SETUP must be run as
.br

\fBSETUP\fP args::<target-test-args>
.br

Only the arguments to the target test are needed, since target
tests do not involve other machines.
.TP
.B
START
.br
Begins running the test which has been configured. This command can only be run
after SETUP.
.TP
.B
STOP
.br
Ends the currently running test. This command can only be run after START, and
only when ipbenchd is running in target mode.
In client mode a test will finish based on completing certain tasks.
In target mode a test will keep running until ordered to stop with this command.

.LP
Once one user sends HELLO to ipbenchd, no other users will be able to send
test-related commands until the current test is completed or aborted.
This will be displayed in STATUS output as "Running a test".
.LP
The state of which of HELLO, LOAD, SETUP, START, STOP should be run next
is maintained in your connection to ipbenchd, not globally. If you disconnect
from ipbenchd, you'll need to prepare your test again from the beginning.
Ipbenchd won't remember what state you reached.
.LP
Outside of debugging, you should never need to run HELLO, LOAD, SETUP, START,
or STOP manually. Use the ipbench controller program to handle running
benchmarks, which will automatically issue these commands in the correct
order with the correct syntax for you.
.LP
The main reason to manually connect to ipbenchd is to check STATUS
or use ABORT to clean up from a benchmark run which failed unexpectedly.
You can also configure ipbench to send ABORT to each client before running
a test.

.SH "FILES"
.LP
(none)
.SH "ENVIRONMENT VARIABLES"
.LP
(none)
.SH "EXAMPLES"
.LP
examples
.SH "AUTHORS"
.LP
ipbenchd by Ian Wienand <ianw@gelato.unsw.edu.au>,
.br
Lesley Rossouw <lesley.rossouw@unsw.edu.au>,
.br
Peter Chubb <peter.chubb@unsw.edu.au>.
.br
Man page by Christopher Irving <c.irving@unsw.edu.au>.
.SH "SEE ALSO"
.LP
ipbench(1)


ChangeLog
2021-09-27  Peter Chubb  <peter.chubb@unsw.edu.au>

	* Update to Python3

2010-03-04  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* Depend on python-lxml to get full XPATH support for ipbench
	  controller. 
	* nfsstone fixes.  Used to dump core in ipbenchd when used.
	

2010-02-10  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* nfs_latency fixes -- now obeys similar arguments to the latency
	test (warmup, cooldown in seconds, samples=)
	* Updated manual.

2010-02-02  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* Fix compilation on Ubuntu Jaunty.

2006-08-28  Ian Wienand  <ianw@localdomain>

	* src/pymod/ipbench.c: fix up some warnings
	* src/pymod/ipbench.i.in: move around so that functions have
	definitions, fixes a few warnings.

2006-03-08  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* debian/ipbench-plugins.install: Don't install libraries as part
	of this package; instead depend on libipbench package (which is
	done automatically).

2006-03-07  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : respect DESTDIR

2005-06-29  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : install libraries in python site-dir

2005-06-23  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/latency/microuptime* : move into 
	* src/libs/* : add microuptime into util library
	* src/tests/latency/latency.[c|h] : remove references to
	microuptime as moved into util library.

2005-06-21  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/nfs_latency/*: start a new nfs latency test, written
	by Luke Macpherson.

2005-05-05  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/ipbench.py, src/pymod/*, src/tests/* : modify API to have a
	controller setup function and phase.  This was we can pass
	arguments to the controller.
	* src/tests/latency.c : add ability to dump out client results to
	a file

2005-03-02  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/latency/latency.c : modify printf specifier for
	tick_rate
	* src/ipbench.py : add more information to execption
	errors including return code and error strings.

2005-02-24  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/cpu_target_lukem/cpu_target_lukem.c : wait on a
	condition variable to support FreeBSD.  also nice() process before
	starting usage thread.

2005-02-22  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* many: changes to bring consisentcy to types; turn on -Wall,
	-Werror; modify to build
	* configure.in: version to 2.0.2

2005-02-18  Mark Tsui <markt@cse.unsw.edu.au>

        * src /tests/latency/microuptime.h:
	* src /tests/cpu_target_lukem/cpu_target_lukem.h: added support for amd64

2005-02-18  Mark Tsui <markt@cse.unsw.edu.au>

	* src/tests/latency.c(latency_marshall): don't use size_t but int

2005-02-18  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/ipbench.c(marshall) : don't pass in uninitalised pointers
	* src/tests/wrapper/{client|target}_wrapper.c: as above
	* src/ipbenchd.py : set reuse addr to stop annoying socket in use problem

2005-01-25  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/cpu_target_lukem/* : add lukem cpu usage test

2004-12-21  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : install using DESTDIR
	* debian/* : debianisation

2004-10-21  Darren Williams  <dsw@quasar>

	* Intoduce a common ipbench2 include directory
	* Add bswap.h to new include directory, code taken
	  from latency.h

2004-10-20  Darren Williams  <dsw@gelato.unsw.edu.au>

	* Import nfsstone benchmark to src/tests/nfsstone.
	* See README in src/tests/nfsstone for details on the nfsstone benchmark.

2004-10-20  Ian Wienand  <ianw@gelato.unsw.edu.au>
	* src/ipbenchd.py : make mmap use -1 as the fd so it is portable
	* src/tests/http/Makefile.am : add to make sure make dist works
	* configure.in : update to make sure all makefile are created for make dist,
	add --enable-tests option to only build some tests.
	* src/tests/Makefile.am : change to follow conditional compliation from --enable-tests
	* src/tests/nfsstone : added by dsw, working in beta stage
	* src/pymod/Makefile.am : do not redefine libdir to install python libraries
	* src/pymod/ipbench.c : use pkglibdir for search path for plugins

2004-09-27  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* restart : restart this ChangeLog for V2


ipbench2/src/ipbenchd.py
#!/usr/bin/env python3
"""
ipbenchd.py

Logic for the ipbench daemon. Deployed on testers and target.
"""

import socketserver
import sys
import re
import os
import signal
import mmap
from argparse import ArgumentParser
from traceback import print_exc

HELP_TEXT = "\nHELP OPTIONS"           \
            "\n------------"           \
            "\nHELP"                   \
            "\nHELLO"                  \
            "\nLOAD"                   \
            "\nSETUP target::<test-target>||port::<test-port>||args::<test-args>" \
            "\nSTART"                  \
            "\nABORT"                  \
            "\nSTATUS"                 \
            "\nQUIT\n"

HELP_TEXT_TARGET = "\nHELP OPTIONS" \
                   "\n------------" \
                   "\nHELP"         \
                   "\nHELLO"        \
                   "\nLOAD"         \
                   "\nSETUP args::<target-test-args>"   \
                   "\nSTART"        \
                   "\nSTOP"         \
                   "\nABORT"        \
                   "\nSTATUS"       \
                   "\nQUIT\n"

OPTIONS = None
SERVER = None
SHUNT = None
TEST_RUNNING = None

# we need something that can be seen by all processes
# this is by no means a real mutex, but it does the job


class TestStatus:
    """
    Class for concurrency control. Creates a shared mmap to move
    information between processes concurrently.
    """

    def __init__(self):
        self.mmap_obj = mmap.mmap(-1, 20,
                                  mmap.MAP_SHARED | mmap.MAP_ANONYMOUS)
        self.clear_pid()

    # we really have to hack around the python concept of a mmap
    def write_pid(self, pid):
        self.mmap_obj.write(str(pid).encode('utf-8') + b'\n')
        self.mmap_obj.seek(0)

    def read_pid(self):
        string_pid = self.mmap_obj.readline()
        self.mmap_obj.seek(0)
        return int(string_pid)

    def clear_pid(self):
        self.mmap_obj[0:2] = b'0\n'
        dbprint("[clear_pid] PID Cleared, now " + str(self.read_pid()))

    def running(self, pid):
        dbprint("[running] setting running pid to " + str(pid))
        self.write_pid(pid)

    def not_running(self):
        self.clear_pid()

    def is_running(self):
        p = self.read_pid()
        return (p > 0)

    def running_pid(self):
        return self.read_pid()


#
# debugging output
#
def dbprint(msg):
    """
    Debug print.

    Param:
        msg: Message to print
    """
    if OPTIONS.debug:
        print(msg, file=sys.stderr)


def debugging():
    """
    Returns true if debug mode triggered.
    """
    return OPTIONS.debug


def str_status(errCode, *msg):
    errCodes = {
        100: "IPBENCH V1.0",
        200: "OK",
        220: "VALID DATA",
        221: "INVALID DATA",
        222: "CLOSING CONTROL CONNECTION",
        400: "ERROR",
        421: "SERVICE NOT AVAILABLE",
        451: "LOCAL PROCESSING ERROR",
        500: "SYNTAX ERROR"
    }

    if len(msg) == 0:
        return repr(errCode) + " " + errCodes[errCode] + "\n"
    return repr(errCode) + " " + errCodes[errCode] + " (" + " ".join(msg) + ")\n"


class StateMachine:
    """
    Class encapsulating a generic state machine for ipbench classes.
    """

    def __init__(self):
        self.handlers = {}
        self.start_state = None
        self.end_states = []

    def add_state(self, name, handler, end_state=0):
        """
        Add a state to the state machine.

        Unless end_state is set, this is assumed to be an intermediary piece of state
        """
        name = name.upper()
        self.handlers[name] = handler
        if end_state:
            self.end_states.append(name)

    def set_start(self, name):
        """
        Set starting state. State must be added first and selected by name.
        """
        self.start_state = name.upper()

    def run(self, cargo):
        """
        Run the state machine.
        """
        try:
            handler = self.handlers[self.start_state]

        # TODO: improve this. Exception handling here is too broad.
        except:
            raise Exception("InitializationError",
                            "must call .set_start() before .run()")

        if not self.end_states:
            raise Exception("InitializationError",
                            "at least one state must be an end_state")

        while True:
            (new_state, cargo) = handler(cargo)
            if new_state.upper() in self.end_states:
                break
            else:
                handler = self.handlers[new_state.upper()]


##
# Client Object
# This implements the state machine
##
class IpbenchClient:
    """
    Client class. This is implemented by the daemon if it is called as a client.

    Implements the state machine class from above.
    """

    def __init__(self, fds):
        global SHUNT, OPTIONS

        dbprint("IpbenchClient [__init__] : client mode")
        import ipbench_client
        SHUNT = ipbench_client

        if OPTIONS.debug:
            SHUNT.enable_debug()

        (self.rfile, self.wfile) = fds

    def send(self, msg):
        """
        Send a message to the attached socket.
        """
        self.wfile.write(msg.encode('utf-8'))

    def readline(self):
        """
        Receive from the attached socket.
        """
        return self.rfile.readline().decode('utf-8')

    def parse_common_opts(self):
        """
        Attempt to parse options common to all tests.
        """
        global TEST_RUNNING, OPTIONS
        while (1):
            line_in = self.readline().strip()

            cmd = re.match(r"\S*", line_in).group(0)
            if cmd is None:
                self.send(str_status(500))
                continue

            elif cmd == "HELP":
                if OPTIONS.target:
                    self.send(HELP_TEXT_TARGET)
                else:
                    self.send(HELP_TEXT)
                self.send(str_status(200))

            elif cmd == "STATUS":
                if OPTIONS.target:
                    self.send("Target Daemon mode\n")
                else:
                    self.send("Testing Daemon mode\n")
                if not TEST_RUNNING.is_running():
                    self.send("Not running a test\n")
                else:
                    pid = TEST_RUNNING.running_pid()
                    self.send("Running a test (%d)\n" % pid)
                self.send(str_status(200))

            else:
                dbprint("[parse_common_opts]: got line " + line_in)
                return line_in

    def listen(self, cargo):
        """
        Client listen loop. Awaits instructions sent by socket and attempts to handle them.
        """
        global TEST_RUNNING
        dbprint("[listen] start")
        while True:
            line_in = self.parse_common_opts()
            if line_in == "ABORT":
                return ("ABORT", 1)
            if line_in == "QUIT":
                return ("ERROR", (222, "Come back soon!"))

            cmd = re.match(r"(\S+)", line_in)
            if cmd is None:
                self.send(str_status(500))
                continue
            else:
                cmd = cmd.group(0)

            if cmd == "HELLO":
                # process the hello. if we are already running, bail out
                if TEST_RUNNING.is_running():
                    return ("ERROR", (421, "Already Running"))
                else:
                    self.send(str_status(200, "Ready to go"))
                    TEST_RUNNING.running(os.getpid())
                    return ("SETUP", 1)

            self.send(str_status(500))

    def setup(self, val):
        """
        Setup the client. Must be called before entering listen loop.
        """
        global SHUNT
        dbprint("[setup] start")
        while True:
            line_in = self.parse_common_opts()
            if line_in == "ABORT":
                return ("ABORT", 1)
            if line_in == "QUIT":
                return ("ERROR", (222, "Come back soon!"))

            rematch = re.match(r"(\w+) ([\s\w-]+)", line_in)
            if rematch is None:
                self.send(str_status(500))
                continue
            else:
                cmd = rematch.group(1)
                arg = rematch.group(2)

                if cmd == "LOAD":
                    try:
                        SHUNT.load_plugin(arg)
                    except RuntimeError as detail:
                        return ("ERROR", (451, str(detail).strip()))
                    else:
                        # Loop until setup commands received.
                        while 1:
                            self.send(str_status(200))

                            line_in = self.parse_common_opts()
                            if line_in == "ABORT":
                                return ("ABORT", 1)
                            if line_in == "QUIT":
                                return ("ERROR", (222, "Come back soon!"))

                            rematch = re.match(r"(\w+) (..*)", line_in)
                            if rematch is None:
                                self.send(str_status(500))
                                continue
                            else:
                                cmd = rematch.group(1)
                                arglist = rematch.group(2)

                                for arg in arglist.split("||"):
                                    (subcmd, val) = arg.split("::")
                                    if subcmd == "target":
                                        test_target = val
                                    if subcmd == "port":
                                        test_port = int(val)
                                    if subcmd == "args":
                                        test_args = val[1:-1]

                            if cmd == "SETUP":
                                try:
                                    if not OPTIONS.target:
                                        dbprint("[setup] Setting up client with " +
                                                test_target + " " + repr(test_port) + " " + test_args)
                                        SHUNT.setup(
                                            test_target, test_port, test_args)
                                    else:
                                        dbprint(
                                            "[setup] Setting up target with args " + test_args)
                                        SHUNT.setup(test_args)

                                except RuntimeError as detail:
                                    return ("ERROR", (451, str(detail).strip()))

                            self.send(str_status(200))
                            return ("RUN", 1)

                self.send(str_status(500))

    def run(self, val):
        """
        Run a test. Expects parameters to be sent over socket.
        """
        dbprint("[run] start")
        while 1:
            line_in = self.parse_common_opts()
            if line_in == "ABORT":
                return ("ABORT", 1)
            if line_in == "QUIT":
                return ("ERROR", (222, "Come back soon!"))

            rematch = re.match(r"(\w+)", line_in)
            if rematch is None:
                self.send(str_status(500))
                continue
            else:
                cmd = rematch.group(1)

                # the start command kicks us off; when we return we go
                # into marshall phase
                if cmd == "START":
                    dbprint("[run] starting actual test")
                    try:
                        SHUNT.start()
                    except RuntimeError as detail:
                        return ("ERROR", (451, str(detail).strip()))
                    dbprint("[run] test complete")
                    return ("MARSHALL", 1)
                self.send(str_status(500))

    def marshall(self, val):
        """
        Marshall data.
        """
        global TEST_RUNNING
        dbprint("[run] marshall start")
        try:
            (valid, data) = SHUNT.marshall()
        except RuntimeError as detail:
            return ("ERROR", (451, str(detail).strip()))

        if valid != 0:
            self.send(str_status(221, "Data to follow"))
        else:
            self.send(str_status(220, "Data to follow"))

        self.send("Content-length: " + repr(len(data)) + "\n")
        self.wfile.write(data)

        TEST_RUNNING.not_running()
        return ("RESET", 1)

    # dummy function that falls out of the state machine and resets
    def reset(self, fd): pass

    def error(self, val):
        global TEST_RUNNING
        dbprint("[error] start")
        (errCode, msg) = val
        # if aborting because test is already running, don't
        # flag test as not running!
        if errCode != 421:
            TEST_RUNNING.not_running()
        self.send(str_status(errCode, msg))
        return ("RESET", 1)

    def abort(self, val):
        global TEST_RUNNING
        if not TEST_RUNNING.is_running():
            dbprint("[abort] test isn't actually running ...")
            return ("RESET", 1)

        pid_to_kill = TEST_RUNNING.running_pid()
        dbprint("[abort] " + repr(os.getpid()) +
                " aborting " + repr(pid_to_kill))
        try:
            os.kill(pid_to_kill, signal.SIGINT)
        except ProcessLookupError:
            TEST_RUNNING.not_running()
            return
        try:
            os.kill(pid_to_kill, signal.SIGKILL)
        except ProcessLookupError:
            pass
        # usually we will have killed *ourself*, but possibly not
        dbprint("[abort] reset")
        return ("RESET", 1)


class IpbenchTarget(IpbenchClient):
    """
    ipbench target. Encapsulates logic for target mode, only entered if
    daemon started with --target param.
    """

    def __init__(self, fds):
        global SHUNT, OPTIONS

        dbprint("[__init__] IpbenchClient [__init__] : target mode")
        import ipbench_target  # SWIG compiled
        SHUNT = ipbench_target

        if OPTIONS.debug:
            SHUNT.enable_debug()

        (self.rfile, self.wfile) = fds

    # only run() needs to be different for the target; as the
    # test must be forked and run in a sub process that is killed
    # when we receive the STOP command
    def run(self, val):
        dbprint("[run] start")
        while 1:
            line_in = self.parse_common_opts()
            if line_in == "ABORT":
                return ("ABORT", 1)
            if line_in == "QUIT":
                return ("ERROR", (222, "Come back soon!"))

            rematch = re.match(r"(\w+)", line_in)
            if rematch is None:
                self.send(str_status(500))
                continue
            else:
                cmd = rematch.group(1)

                if cmd == "START":
                    dbprint("[run] starting target test")
                    ppid = os.getpid()
                    pid = os.fork()
                    if pid > 0:
                        # Parent
                        try:
                            SHUNT.start()
                        except RuntimeError as detail:
                            os.kill(pid, signal.SIGKILL)
                            dbprint("[run] target test start() failed!")
                            return ("ERROR", (451, str(detail).strip()))
                    else:
                        # we are in a child, listen for stop and send SIGUSR1
                        # to parent to stop the test
                        # write_status(self.wfile, 200)
                        dbprint("Child")
                        # self.send("Child")
                        line_in = self.parse_common_opts()
                        if line_in == "ABORT":
                            os.kill(ppid, signal.SIGINT)
                            os._exit(0)
                        if line_in == "QUIT":
                            return ("ERROR", (222, "Come back soon!"))
                        rematch = re.match(r"(\w+)", line_in)
                        if rematch is None:
                            self.send(str_status(500))
                            continue
                        else:
                            cmd = rematch.group(1)
                            if cmd == "STOP":
                                dbprint(
                                    "[run] got stop command, sending SIGUSR1 to " + repr(ppid))
                                os.kill(ppid, signal.SIGUSR1)
                                # Can't use sys.exit() as it'll close the socket
                                os._exit(0)

                    # In parent
                    dbprint("[run] test complete")
                    return ("MARSHALL", 1)
                # unknown command
                self.send(str_status(500))


class IpbenchRequestHandler(socketserver.StreamRequestHandler):
    """
    Request handler for client and target.
    """

    def handle(self):
        global OPTIONS
        try:
            self.wfile.write(str_status(100).encode('utf-8'))
        except Exception as e:
            dbprint("Exception %s" % str(e))

        try:
            dbprint("[handle_child] new handler ["+repr(os.getpid())+"]")

            m = StateMachine()
            if OPTIONS.target:
                c = IpbenchTarget((self.rfile, self.wfile))
            else:
                c = IpbenchClient((self.rfile, self.wfile))

            # setup state machine
            m.add_state("LISTEN", c.listen)
            m.add_state("SETUP", c.setup)
            m.add_state("RUN", c.run)
            m.add_state("MARSHALL", c.marshall)
            m.add_state("ERROR", c.error)
            m.add_state("ABORT", c.abort)
            m.set_start("LISTEN")
            m.add_state("RESET", c.reset, end_state=1)
            m.run(1)

            dbprint("[handle] state machine done")
        except Exception as e:
            # exit and close
            dbprint("[handle] " + repr(os.getpid()) +
                    " caught exception ... finishing up: " + str(e))
            print_exc()
            sys.exit(0)


class IpbenchClientServer(socketserver.ForkingMixIn, socketserver.TCPServer):
    """
    Class to extend the socket server to replace error handling.
    """
    allow_reuse_address = True

    def handle_error(self, request, client_address):
        """
        Handle an error gracefully.  Overrides standard method.

        The default is to print a traceback and continue, except in
        the case of a SystemExit exception, which should terminate
        processing.
        """
        # TODO: replace this. this code clearly does not work in python3
        raise
        dbprint("[handle_error] handling it ...")
        import exceptions
        etype, evalue = sys.exc_info()[:2]
        if etype is exceptions.SystemExit:
            raise
# extra exception debugging if you need it
#        print '-'*40
#        print 'Exception happened during processing of request from',
#        print client_address
#        import traceback
#        traceback.print_exc(1000, sys.stdout)
#        print '-'*40
#        sys.stdout.flush()


def main():
    """
    ipbench daemon main function.
    """
    global OPTIONS, SERVER
    parser = ArgumentParser(description="ipbenchd can act as a client or a target in the ipbench benchmarking suite.")
    parser.add_argument("--version", action="version", version="%(prog)s 2.1.1")
    parser.add_argument("-i", "--ip", dest="ip",
                      help="Makes the daemon bind to the specified ip. Default is to bind to all available ip addresses.", type=str, default="", action="store")
    parser.add_argument("-p", "--port", dest="port",
                      help="Makes the daemon listen on the specified port. Default is 8036 for clients, 8037 for targets.", type=int, default=8036, action="store")
    parser.add_argument("-d", "--debug", dest="debug", action="store_true",
                      help="Enable debugging; provide verbose output.", default=False)
    parser.add_argument("-t", "--target", dest="target", action="store_true",
                      help="Put the daemon in target mode.")

    OPTIONS = parser.parse_args()

    if OPTIONS.target:
        OPTIONS.port = 8037

    SERVER = IpbenchClientServer(
        (OPTIONS.ip, OPTIONS.port), IpbenchRequestHandler)

    global TEST_RUNNING
    TEST_RUNNING = TestStatus()
    dbprint("[main] listening on port " +
            repr(OPTIONS.port) + " [" + repr(os.getpid()) + "]")

    try:
        SERVER.serve_forever()
    except:
        dbprint("[main] caught exception from serve_forever()")


if __name__ == "__main__":
    main()
